### PR TITLE
Service node checkpointing: rebased on relaxed deregistration

### PR DESCRIPTION
**Pull request by Doy-lee**
This gets checkpointing back to feature parity from my ServiceNodeCheckpointing5 branch, i.e. fixing some synchronisation bugs that deleted checkpoints prematurely, adding checkpoints in the add_new_block step instead of wholesale in protocol handler stage.

Still no deregistration, this is all just manual merging what I had and using the new things setup in the relaxed registration.

Also fixes the unit_tests/core_tests that were not working and pop_blocks not calling the detached hooks persisting checkpoints after a reorg.

### CODE REVIEW

File: src/cryptonote_core/service_node_quorum_cop.cpp [hunk 1]
```diff
@@ -118,6 +118,10 @@ namespace service_nodes
 
   void quorum_cop::process_quorums(cryptonote::block const &block)
   {
+    int const hf_version = block.major_version;
```
Comment (msgmaxim) [inline src/cryptonote_core/service_node_quorum_cop.cpp]: `uint8_t`/`auto`

File: src/cryptonote_core/cryptonote_core.h [hunk 2]
```diff
@@ -780,7 +780,7 @@ namespace cryptonote
       *
       * @param type The quorum type to retrieve
       * @param height Block height to deterministically recreate the quorum list from
-
+const
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/cryptonote_core/cryptonote_core.h]: oops. fixed

File: src/p2p/net_node.inl [hunk 3]
```diff
@@ -2009,7 +2009,7 @@ namespace nodetool
       return 1;
     }
 
-#if !defined(LOKI_ENABLE_INTEGRATION_TEST_HOOKS)
+#if 0
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/p2p/net_node.inl]: Yeah this one should be reverted. Fixed

File: src/cryptonote_protocol/cryptonote_protocol_handler.inl [hunk 4]
```diff
@@ -1333,12 +1330,63 @@ namespace cryptonote
...
+              {
+                MERROR("Checkpoint blob given but not expecting a checkpoint at this height");
+                return false;
+              }
+
+              // TODO(doyle): If we are receiving alternative blocks, we won't
+              // have the quorum for the alternative chain meaning we will not
+              // be able to verify the checkpoint. For now always accept
+              // whatever checkpoint we receive
+#if 0
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/cryptonote_protocol/cryptonote_protocol_handler.inl]: This one intentional unfortunately not going to be in the coming release but later down the line

File: src/cryptonote_core/tx_pool.cpp [hunk 5]
```diff
@@ -99,13 +99,17 @@ namespace cryptonote
...
     public:
-      LockedTXN(Blockchain &b): m_blockchain(b), m_batch(false) {
+      LockedTXN(Blockchain &b): m_blockchain(b), m_batch(false), m_active(false) {
         m_batch = m_blockchain.get_db().batch_start();
+        m_active = true;
       }
-      ~LockedTXN() { try { if (m_batch) { m_blockchain.get_db().batch_stop(); } } catch (const std::exception &e) { MWARNING("LockedTXN dtor filtering exception: " << e.what()); } }
+      void commit() { try { if (m_batch && m_active) { m_blockchain.get_db().batch_stop(); m_active = false; } } catch (const std::exception &e) { MWARNING("LockedTXN::commit filtering exception: " << e.what()); } }
+      void abort() { try { if (m_batch && m_active) { m_blockchain.get_db().batch_abort(); m_active = false; } } catch (const std::exception &e) { MWARNING("LockedTXN::abort filtering exception: " << e.what()); } }
+      ~LockedTXN() { abort(); }
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/cryptonote_core/tx_pool.cpp]: Updated

File: src/cryptonote_core/service_node_rules.h [hunk 6]
```diff
@@ -39,15 +37,17 @@ namespace service_nodes {
...
-  constexpr uint64_t CHECKPOINT_INTERVAL                       = 4;  // Checkpoint every 4 blocks and prune when too old except if (height % CHECKPOINT_STORE_PERSISTENTLY_INTERVAL == 0)
-  constexpr uint64_t CHECKPOINT_STORE_PERSISTENTLY_INTERVAL    = 60; // Persistently store the checkpoints at these intervals
-  constexpr uint64_t CHECKPOINT_VOTE_LIFETIME                  = CHECKPOINT_STORE_PERSISTENTLY_INTERVAL; // Keep the last 60 blocks worth of votes
-#if defined(LOKI_ENABLE_INTEGRATION_TEST_HOOKS)
-  constexpr size_t   CHECKPOINT_QUORUM_SIZE                    = 1;
-  constexpr size_t   CHECKPOINT_MIN_VOTES                      = 1;
+  constexpr uint64_t  CHECKPOINT_INTERVAL                    = 4;  // Checkpoint every 4 blocks and prune when too old except if (height % CHECKPOINT_STORE_PERSISTENTLY_INTERVAL == 0)
+  constexpr uint64_t  CHECKPOINT_STORE_PERSISTENTLY_INTERVAL = 60; // Persistently store the checkpoints at these intervals
+  constexpr uint64_t  CHECKPOINT_VOTE_LIFETIME               = CHECKPOINT_STORE_PERSISTENTLY_INTERVAL; // Keep the last 60 blocks worth of votes
+#if 1
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/cryptonote_core/service_node_rules.h]: Yes, ifdef-ed around the integration test hooks #define

File: src/cryptonote_core/service_node_voting.cpp [hunk 7]
```diff
@@ -214,6 +242,53 @@ namespace service_nodes
...
+        LOG_PRINT_L1("Checkpoint has too many signatures to be considered");
+        return false;
+      }
+
+      std::array<size_t, service_nodes::CHECKPOINT_QUORUM_SIZE> unique_vote_set = {};
+      for (service_nodes::voter_to_signature const &voter_to_signature : checkpoint.signatures)
+      {
+        if (!bounds_check_worker_index(quorum, voter_to_signature.voter_index, nullptr)) return false;
+
+        if (++unique_vote_set[voter_to_signature.voter_index])
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/cryptonote_core/service_node_voting.cpp]: Yep, bug. Fixed

File: src/cryptonote_core/service_node_quorum_cop.cpp (hunk 1 above)
Comment (Doy-lee) [reply src/cryptonote_core/service_node_quorum_cop.cpp]: Updated

**Pull request merged**